### ISSUE TITLE
Will constrained optimization be supported?

### ISSUE THREAD

**Issue by zhouxian** [patrick-kidger/optimistix#45]
Hi, thank you for the amazing library! 

I was wondering if minimize with user-specified bounds, or algorithms like projected gradient descent be supported?
If not, what would be the best practice you suggest if we are trying to solve `argmin(norm(Ax+b)) s.t. x >=0`?

Thanks in advance!

**Comment by patrick-kidger** [comment]
So right now we do support hypercube constraints in our root-finding methods (e.g. `Newton` or `Bisection`). And moreover it is sort-of-possible to cast a minimisation problem as a root-finding problem, by seeking a root of the gradient of the function. (You have to be careful as this may also converge to any other external point.) So I think the particular example you've given could be tackled today, although I'm not sure how efficient it would be as compared to a dedicated solver.

Other than that we don't currently support any kind of constraints in our solves. We may do at some point, but as always developer time is the limiting factor!

**Comment by mjo22** [comment]
Hello! Was also curious on this front. In particular, I am wondering if there are plans to support box constraints in BFGS, or if this can already be done in the current framework. `jaxopt` seems to have an implementation of this, but I’m not sure how well maintained it is.

**Comment by patrick-kidger** [comment]
For something simple like box constraints, then it should be fairly easy to add that in. We'd be happy to take a PR adding that. (Note that box constraints are already implemented for the root-finders, so you could follow a similar pattern to there.)

**Comment by f0uriest** [comment]
If box constraints are supported it shouldn't be too much work to add an augmented Lagrangian type method for general constraints. Could maybe be like a "meta solver" where you can wrap another solver to manage the unconstrained sub problems? I can hopefully take a look at it once I finish up a few other projects.

**Comment by billtubbs** [comment]
Constrained nonlinear optimization would be great.

I'm not an optimization expert, so this may be a lot to ask, but what would it take to get a solver like [IPOPT](https://github.com/coin-or/Ipopt/tree/releases/3.14.16) integrated with Jax?  IPOPT is a tried and tested solver for large scale nonlinear optimization (with nonlinear constraints) used widely in engineering fields like chemical engineering, control, and robotics (see for example [CasADi](https://web.casadi.org/), [GEKKO](https://gekko.readthedocs.io/en/latest/), [Pyomo](https://www.pyomo.org/)).  Would be great to see more people in these fields trying out Jax.

From my limited understanding, solving large scale problems requires the solver to take advantage of the sparsity of the constraint Jacobian and Hessian matrices.  Is this a fundamental problem in Jax?

**Comment by johannahaffner** [comment]
Hi @billtubbs, 

since box constraints (mentioned above) are just a special case of inequality constraints, I have been looking into ways to make some of the "ingredients" used for their implementation in various methods available. It is still early days though! 

Thank you for the solver suggestions, I will look into them :)

**Comment by f0uriest** [comment]
If I remember correctly, IPOPT uses a filter to decide on accepting steps, which might be difficult to implement in JAX with compile time memory bounds? It can also switch between different "modes" dynamically which might make `vmap` super inefficient. 

Something like LANCELOT/GALAHAD or SLSQP would probably be easier.

As for sparsity, that is usually what lets you solve large problems efficiently, though you don't necessarily need actual sparse data structures. It should be easy to just use some of the iterative solvers from `lineax` on the KKT system, though you may need a good preconditioner. @patrick-kidger is it possible to pass linear solver options like preconditoners through `optimistix` to `lineax`?

**Comment by patrick-kidger** [comment]
I think frequently "no" but happy to consider a PR changing that fact :)

**Comment by billtubbs** [comment]
> If I remember correctly, IPOPT uses a filter to decide on accepting steps, which might be difficult to implement in JAX with compile time memory bounds? It can also switch between different "modes" dynamically which might make `vmap` super inefficient.
> 
> Something like LANCELOT/GALAHAD or SLSQP would probably be easier.
> 
> As for sparsity, that is usually what lets you solve large problems efficiently, though you don't necessarily need actual sparse data structures. It should be easy to just use some of the iterative solvers from `lineax` on the KKT system, though you may need a good preconditioner. @patrick-kidger is it possible to pass linear solver options like preconditoners through `optimistix` to `lineax`?

I think you're right, it would be impossible to differentiate through an optimizer like IPOPT.  Actually, all I was thinking was to use IPOPT with Jax. In other words, IPOPT would use Jax to efficiently compute the constraint Jacobian etc. But sounds like the lack of a sparse matrix representation might make that difficult.

**Comment by billtubbs** [comment]
Related information I found:
- [Can we use the Jax library with ipopt to calculate the hessian matrix and Jacobian and still be able to define the structures?](https://or.stackexchange.com/questions/9596/can-we-use-the-jax-library-with-ipopt-to-calculate-the-hessian-matrix-and-jacobi)
- [feature request: sparse jacobian and sparse hessians](https://github.com/jax-ml/jax/issues/1032)

**Comment by johannahaffner** [comment]
> I think you're right, it would be impossible to differentiate through an optimizer like IPOPT. 

In what context would you be interested in this? Do you need to obtain a derivative of your function with respect to the solution of the optimisation problem?

> Actually, all I was thinking was to use IPOPT with Jax. In other words, IPOPT would use Jax to efficiently compute the constraint Jacobian etc. But sounds like the lack of a sparse matrix representation might make that difficult.

From the IPOPT documentation, [I got this](https://coin-or.github.io/Ipopt/INTERFACES.html) (for triplet format matrix specification)

> Ipopt needs to know the number of nonzero elements and the sparsity structure (row and column indices [...]) of the constraint Jacobian and the Lagrangian function Hessian. [...] this nonzero structure MUST remain constant [...].

I did some rudimentary benchmarking for a function with a diagonal Hessian and one with simple coupling terms on the off-diagonal, adding a `jnp.take` before returning select elements. It does seem as though XLA is able to optimise a little in this special case. I get speed-ups of about 2,5x for $n = 10000$, and Hessian $\in \mathbb{R}^{n \times n}$.

---

I'm actually working on an interior point method right now! In optimistix :) 

Given that I have you here: do you have an idea what parts of IPOPT contribute strongly to its performance? I took a good look at their implementation paper a few weeks ago. I think a lot of these things could in principle be done in JAX, although some might just not be a good fit. 

For example, the filter is re-set regularly, and may also be bypassed under some conditions, because keeping a history of iterates is not always helpful, even if it often is. Do you have an idea how often these things are invoked on your real data?

**Comment by billtubbs** [comment]
From a big picture perspective, the argument for a convergence between modern machine learning tools and approaches and "classical" engineering or model-based optimization is quite compelling so I think I'm not the only one who would like to see more engineers using Jax.

Not sure if you're seen it but this talk by Jan Drogna provides a nice big picture perspective on the current tools and their strengths and weaknesses:

<img width="1233" alt="Screen Shot 2025-01-08 at 10 41 59 " src="https://github.com/user-attachments/assets/26954700-e074-453b-96f9-c64b08a80e7f" />

https://www.youtube.com/watch?v=z_01ZQ2DskA

**Comment by johannahaffner** [comment]
Fully agree on the big picture! There are so many other fields that can now benefit from all the innovations in modern scientific computing that were necessary to get ML to where it is today.

> provide an interface to the original IPOPT solver in the same way that it is used in other libraries such as CasADi, GEKKO.

That sounds reasonable - although I think optimistix would not be the right place for this. `IPOPT` supports interfacing through TNLP - this entails defining an optimisation problem (as an object) that provides methods for the evaluation of the target function and gradients. This is very different to how optimistix works, where you only provide the function, and how the derivatives are computed is then solver-specific.

---
> Hope that helps, and I appreciate your willingness to consider this.

Thank you for letting me know more about the problems you're interested in solving! The multiple shooting example is really cool. Sorry I can't be of more help right now, though!
